sw/device/silicon_creator/mask_rom/primitive_bootstrap.c

```diff
@@ -0,0 +1,218 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/mask_rom/primitive_bootstrap.h"
+
+#include <stddef.h>
+
+#include "sw/device/boot_rom/spiflash_frame.h"
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_gpio.h"
+#include "sw/device/lib/dif/dif_spi_device.h"
+#include "sw/device/lib/flash_ctrl.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/silicon_creator/lib/drivers/hmac.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+#define GPIO_BOOTSTRAP_BIT_MASK 0x00020000u
+
+static dif_spi_device_t spi;
+static dif_spi_device_config_t spi_config;
+
+static rom_error_t spi_device_init(void) {
+  if (dif_spi_device_init(
+          mmio_region_from_addr(TOP_EARLGREY_SPI_DEVICE_BASE_ADDR), &spi) !=
+      kDifOk) {
+    return kErrorBootstrapSpiDevice;
+  }
+
+  spi_config.clock_polarity = kDifSpiDeviceEdgePositive;
+  spi_config.data_phase = kDifSpiDeviceEdgeNegative;
+  spi_config.tx_order = kDifSpiDeviceBitOrderMsbToLsb;
+  spi_config.rx_order = kDifSpiDeviceBitOrderMsbToLsb;
+  spi_config.rx_fifo_timeout = 63;
+  spi_config.rx_fifo_len = kDifSpiDeviceBufferLen / 2;
+  spi_config.tx_fifo_len = kDifSpiDeviceBufferLen / 2;
+
+  if (dif_spi_device_configure(&spi, &spi_config) != kDifOk) {
+    return kErrorBootstrapSpiDevice;
+  }
+  return kErrorOk;
+}
+
+static rom_error_t spi_device_rx_pending(size_t *bytes_available) {
+  if (dif_spi_device_rx_pending(&spi, &spi_config, bytes_available) != kDifOk) {
+    return kErrorBootstrapSpiDevice;
+  }
+  return kErrorOk;
+}
+
+static rom_error_t spi_device_recv(void *buf, size_t buf_len) {
+  if (dif_spi_device_recv(&spi, &spi_config, buf, buf_len,
+                          /*bytes_received=*/NULL) != kDifOk) {
+    return kErrorBootstrapSpiDevice;
+  }
+  return kErrorOk;
+}
+
+static rom_error_t spi_device_send(const void *buf, size_t buf_len) {
+  if (dif_spi_device_send(&spi, &spi_config, buf, buf_len,
+                          /*bytes_received=*/NULL) != kDifOk) {
+    return kErrorBootstrapSpiDevice;
+  }
+  return kErrorOk;
+}
+
+/**
+ * Check if flash is blank to determine if bootstrap is needed.
+ */
+static rom_error_t bootstrap_requested(bool *result) {
+  // Initialize GPIO device.
+  dif_gpio_t gpio;
+  if (
+      dif_gpio_init(mmio_region_from_addr(TOP_EARLGREY_GPIO_BASE_ADDR), &gpio) != kDifOk) {
+        return kErrorBootstrapGpio;
+      }
```

Comment:
it looks like you need to run clang-format for this file ([log](https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=54679&view=logs&j=3a559e2a-952e-58d2-b8db-2e604a9266d7&t=a1c8586a-23d4-5717-ec4f-837a6ab6dd66&l=21))